src/macros.jl

```diff
@@ -811,11 +811,59 @@ esc_nonconstant(x::Number) = x
 esc_nonconstant(x::Expr) = isexpr(x,:quote) ? x : esc(x)
 esc_nonconstant(x) = esc(x)
 
+# Returns the type of what `constructvariable!` would return with the same positional arguments.
+variabletype(m::Model) = Variable
+# Returns a new variable belonging to the model `m`. Additional positional arguments can be used to dispatch the call to a different method.
+# The return type should only depends on the positional arguments for `variabletype` to make sense.
+function constructvariable!(m::Model; _error::Function=error, lowerbound::Number=-Inf, upperbound::Number=Inf, category::Symbol=:Default, objective::Number=0., inconstraints::Vector=[], coefficients::Vector{Float64}=Float64[], basename::AbstractString="", start::Number=NaN, extra_kwargs...)
+    for (kwarg, _) in extra_kwargs
+        _error("Unrecognized keyword argument $kwarg")
+    end
+    if category == :Default
+        category = :Cont
+    end
+    if objective == 0 && isempty(inconstraints)
+        Variable(m, lowerbound, upperbound, category, basename, start)
+    else
+        Variable(m, lowerbound, upperbound, category, objective, inconstraints, coefficients, basename, start)
+    end
+end
+
+variabletype(m::Model, t::Union{Type{Val{:Cont}}, Type{Val{:Int}}, Type{Val{:Bin}}, Type{Val{:SemiCont}}, Type{Val{:SemiInt}}}) = variabletype(m)
+function constructvariable!(m::Model, t::Union{Type{Val{:Cont}}, Type{Val{:Int}}, Type{Val{:Bin}}, Type{Val{:SemiCont}}, Type{Val{:SemiInt}}}; category::Symbol=:Default, extra_kwargs...)
+    cat = t.parameters[1]
+    if category != :Default && category != cat
+        _error("A variable cannot be both of category $cat and $category. Please specify only one category.")
+    end
+    constructvariable!(m; category=cat, extra_kwargs...)
+end
+
+variabletype(m::Model, t::Symbol) = variabletype(m, Val{t})
+function constructvariable!(m::Model, t::Symbol; extra_kwargs...)
+    constructvariable!(m, Val{t}; extra_kwargs...)
```

Comment:
Then we might as well modify `category` directly. The whole point of this method is to support specifying a `Symbol` in `extra` but I start to think that this is not a good idea. In PolyJuMP for example since I will create the `Poly` type I might just as well ask to give `Poly` instead of `:Poly`.